File: workflow/scripts/prepare-nh-input.R
#!/usr/bin/env Rscript

library(tidyverse)
library(lubridate)
library(arrow)
library(ncdf4)
library(RNetCDF)
library(rnrfa)
library(RcppRoll)
library(yaml)

options(dplyr.summarise.inform = FALSE)

## Extract configuration info
if (sys.nframe() == 0L) {
  args = commandArgs(trailingOnly=TRUE)
  config <- read_yaml(args[1])
  aggr_period = args[2]
  outputroot = args[3]
  args = commandArgs()
  m <- regexpr("(?<=^--file=).+", args, perl=TRUE)
  cwd <- dirname(regmatches(args, m))
}
source(file.path(cwd, "external/R/utils.R")) # TODO eventually put utils in package
config = parse_config(config)

## Create output directories
outputdir <- file.path(outputroot, "analysis", aggr_period, "nh-input") #, "time_series")
dir.create(outputdir, recursive = TRUE)
dir.create(file.path(outputdir, "time_series"))
dir.create(file.path(outputdir, "attributes"))

## ####################################################### ##
## 1 - Time series data
## ####################################################### ##

ds <- open_dataset(file.path(outputroot, "analysis", aggr_period, "input")) %>% collect()
station_ids <- ds$ID %>% unique()
n_stations <- length(station_ids)
## subsets <- x$subset %>% unique()

for (i in 1:n_stations) {
  id <- station_ids[i]
  xx <- ds %>% filter(ID %in% id & subset %in% "best_n")
  yrs <- xx$year + xx$lead_time - 1
  dates <- paste(yrs, "12-01", sep = "-") %>% as.POSIXct(tz = "UTC", format = "%Y-%m-%d")
  tunits <- "days since 1970-01-01"
  nc_dates <- RNetCDF::utinvcal.nc(tunits, dates)

  ## Define dimensions
  ## From neuralhydrology docs: "The netcdf has to have one
  ## coordinate called date, containing the datetime index."
  timedim <- ncdim_def("date", tunits, as.double(nc_dates))

  ## Define variables
  fillvalue <- 1e32
  dlname <- "Q95"
  q95_def <- ncvar_def("Q95", "m3 d-1", timedim, fillvalue, dlname)
  dlname <- "North Atlantic Oscillation index"
  nao_def <- ncvar_def("NAO", "hPa", timedim, fillvalue, dlname)
  dlname <- "East Atlantic index"
  ea_def <- ncvar_def("EA", "hPa", timedim, fillvalue, dlname)
  dlname <- "Atlantic Multidecadal Variability"
  amv_def <- ncvar_def("AMV", "K", timedim, fillvalue, dlname)
  dlname <- "European precipitation"
  prec_def <- ncvar_def("P", "mm d-1", timedim, fillvalue, dlname)
  dlname <- "UK temperature"
  tmp_def <- ncvar_def("T", "K", timedim, fillvalue, dlname)

  ## Create netCDF4 file and put data
  nc_filename <- file.path(outputdir, "time_series", paste0(id, ".nc"))
  ncout <- nc_create(nc_filename, list(q95_def, nao_def, ea_def, amv_def, prec_def, tmp_def), force_v4 = TRUE)
  ncvar_put(ncout, q95_def, xx$Q_95)
  ncvar_put(ncout, nao_def, xx$nao)
  ncvar_put(ncout, ea_def, xx$ea)
  ncvar_put(ncout, amv_def, xx$amv)
  ncvar_put(ncout, prec_def, xx$european_precip)
  ncvar_put(ncout, tmp_def, xx$uk_temp)
  nc_close(ncout)
}

## ####################################################### ##
## 2 - Static attributes
## ####################################################### ##

camels_attr_fs <- c("CAMELS_GB_climatic_attributes.csv",
                    "CAMELS_GB_humaninfluence_attributes.csv",
                    "CAMELS_GB_hydrogeology_attributes.csv",
                    "CAMELS_GB_hydrologic_attributes.csv",
                    ## "CAMELS_GB_hydrometry_attributes.csv",
                    "CAMELS_GB_landcover_attributes.csv",
                    "CAMELS_GB_soil_attributes.csv",
                    "CAMELS_GB_topographic_attributes.csv")

camels_datadir <- file.path(config$aux_data$camels, "data")
destdir <- file.path(outputdir, "attributes")
dir.create(destdir, recursive = TRUE)
for (i in 1:length(camels_attr_fs)) {
  filepath <- file.path(camels_datadir, camels_attr_fs[i])
  file.copy(filepath, destdir, overwrite = TRUE)
}


## ####################################################### ##
## 3 - Basin list
## ####################################################### ##

conn <- file(file.path(outputdir, "basins.txt"))
writeLines(as.character(station_ids[1:5]), conn)
close(conn)

## ## ####################################################### ##
## ## 4 - YAML configuration file
## ## ####################################################### ##

## template <- read_yaml("resources/nh-config-template.yml")
## template$experiment_name <- paste0("cudalstm_", n_stations, "_basins_", aggr_period)
## template$train_basin_file <- "basins.txt"
## template$validation_basin_file <- "basins.txt"
## template$test_basin_file <- "basins.txt"
## template$train_start_date <- "01/12/1961"
## template$train_end_date <- "01/12/2006"
## template$validation_start_date <- "01/12/1961"
## template$validation_end_date <- "01/12/2006"
## template$test_start_date <- "01/12/1961"
## template$test_end_date <- "01/12/2006"
## ## template$device <- "cuda:0"
## template$device <- "cpu"
## template$validate_every <- as.integer(3)
## template$validate_n_random_basins <- as.integer(1)
## template$metrics <- "NSE"
## template$model <- "cudalstm"
## template$head <- "regression"
## template$output_activation <- "linear"
## template$hidden_size <- as.integer(20)
## template$initial_forget_bias <- as.integer(3)
## template$output_dropout <- 0.4
## template$optimizer <- "Adam"
## template$loss <- "MSE"
## template$learning_rate <- "-0: 1e-2\n-30: 5e-3\n-40: 1e-3" #list("0" = "1e-2", "30"= "5e-3", "40"= "1e-3")
## ## template$learning_rate <- list("0" = "1e-2", "30" = "5e-3", "40" = "1e-3")
## template$batch_size <- as.integer(256)
## template$epochs <- as.integer(50)
## template$clip_gradient_norm <- as.integer(1)
## template$predict_last_n <- as.integer(1)
## template$seq_length <- as.integer(8)
## template$num_workers <- as.integer(8)
## template$log_interval <- as.integer(5)
## template$log_tensorboard <- as.integer(0) # FIXME should be True or False, but can't remove quotes [0/1 should be correctly interpreted by Python yaml loader]
## template$log_n_figures <- as.integer(0)
## template$save_weights_every <- as.integer(1)
## template$dataset <- "generic"
## template$data_dir <- noquote(".")
## template$dynamic_inputs <- c("NAO", "EA", "AMV", "P", "T")
## template$target_variables <- "Q95"
## template$clip_targets_to_zero <- "Q95"

## write_yaml(template, file.path(outputdir, "basins.yml"), fileEncoding = "UTF-8")

## test <- read_yaml(file.path(outputdir, "basins.yml"))
## current <- read_yaml("basin.yml")

## items <- names(test)
## for (i in 1:length(items)) {
##   nm <- items[i]
##   x <- test[[nm]]
##   y <- current[[nm]]
##   if (!isTRUE(all.equal(x, y))) {
##     cat(sprintf("Mismatch for item: %s\n", nm))
##     cat(sprintf("New: %s\n", x))
##     cat(sprintf("Ref: %s\n", y))
##   }
## }

File: workflow/scripts/prepare-nh-input.py
#!/usr/bin/env python3

import os
import sys
import shutil
import datetime
from pathlib import Path
from tqdm import tqdm

import pyarrow
import pandas as pd
import xarray as xr
from ruamel.yaml import YAML

# Get command line arguments
config = sys.argv[1]
aggr_period = sys.argv[2]
outputroot = sys.argv[3]

# # For testing:
# config = 'config/config.yml'
# outputroot = 'results'
# aggr_period = 'yr2to9_lag'

outputdir = os.path.join(os.getcwd(), outputroot, 'analysis', aggr_period, 'nh-input')
try:
    os.makedirs(outputdir)
except FileExistsError:
    pass

try:
    os.makedirs(os.path.join(os.getcwd(), outputdir, 'time_series'))
except FileExistsError:
    pass

try:
    os.makedirs(os.path.join(os.getcwd(), outputdir, 'attributes'))
except FileExistsError:
    pass

rundir = os.path.join(os.getcwd(), outputroot, 'analysis', aggr_period, 'nh-output', 'runs')
try:
    os.makedirs(rundir)
except FileExistsError:
    pass

# ##################################### #
# 1 - Catchment attributes              #
# ##################################### #

yaml = YAML() #typ = 'safe')
cfg = yaml.load(Path(config))
camels_datadir = os.path.join(cfg['aux_data']['camels'], 'data')
destdir = os.path.join(outputdir, 'attributes')

camels_attr_fs = ['CAMELS_GB_climatic_attributes.csv',
                  'CAMELS_GB_humaninfluence_attributes.csv',
                  'CAMELS_GB_hydrogeology_attributes.csv',
                  'CAMELS_GB_hydrologic_attributes.csv',
                  'CAMELS_GB_hydrometry_attributes.csv',
                  'CAMELS_GB_landcover_attributes.csv',
                  'CAMELS_GB_soil_attributes.csv',
                  'CAMELS_GB_topographic_attributes.csv']
camels_attr_fs = [os.path.join(camels_datadir, f) for f in camels_attr_fs]

for f in camels_attr_fs:
    srcfile = os.path.join(camels_datadir, f)
    shutil.copy(srcfile, destdir)

# ##################################### #
# 2 - Time series data                  #
# ##################################### #

nrfa_meta = pd.read_parquet(os.path.join(outputroot, 'nrfa-metadata.parquet'))
nrfa_meta = nrfa_meta.astype({'id' : int})
nrfa_meta = nrfa_meta.set_index('id')

x = pd.read_parquet(os.path.join(outputroot, 'analysis', aggr_period, 'input'))
x = x.astype({'year' : int, 'lead_time' : int, 'ID' : int})
station_ids = list(x['ID'].unique())
n_stations = len(station_ids)

for i in tqdm(range(n_stations)):
    stn = station_ids[i]
    xx = x.loc[(x['ID'] == stn) & (x['subset'] == 'best_n')].copy(deep = True)
    fcst_start_years = list(xx['year'] + xx['lead_time'] - 1)
    fcst_start_dates = pd.to_datetime([datetime.datetime(yr, 12, 1, 0, 0) for yr in fcst_start_years])
    xx.loc[:,'date'] = fcst_start_dates
    xx = xx.set_index('date')
    xx = xx[['Q_95', 'nao', 'ea', 'amv', 'european_precip', 'uk_temp']]
    xx = xx.rename(columns = {
        'Q_95' : 'Q95',
        'nao' : 'NAO',
        'ea' : 'EA',
        'amv' : 'AMV',
        'european_precip' : 'P',
        'uk_temp' : 'T'
    })
    # Normalize by catchment area (take step by step for clarity)
    area = nrfa_meta.loc[stn, 'catchment_area'] * 1000 * 1000 # km2 to m2
    xx['Q95'] *= (24 * 60 * 60)                              # m3 s-1 to m3 d-1
    xx['Q95'] /= area                                        # m3 d-1 to m d-1
    xx['Q95'] *= 1000                                        # m d-1 to mm d-1
    xarr = xr.Dataset.from_dataframe(xx)
    nc_filename = os.path.join(outputdir, 'time_series', str(stn) + '.nc')
    xarr.to_netcdf(nc_filename)

# ##################################### #
# 3 - NH configuration                  #
# ##################################### #
climatic_attributes = ['p_mean', 'pet_mean', 'aridity', 'p_seasonality', 'frac_snow', 'high_prec_freq', 'high_prec_dur', 'low_prec_freq', 'low_prec_dur']
human_influence_attributes = []
hydrogeology_attributes = ['inter_high_perc', 'inter_mod_perc', 'inter_low_perc', 'frac_high_perc', 'frac_mod_perc', 'frac_low_perc', 'no_gw_perc', 'low_nsig_perc', 'nsig_low_perc']
hydrologic_attributes = []
landcover_attributes = ['dwood_perc', 'ewood_perc', 'urban_perc']
soil_attributes = ['sand_perc', 'silt_perc', 'clay_perc', 'porosity_hypres', 'conductivity_hypres', 'soil_depth_pelletier_50']
topographic_attributes = ['gauge_lat', 'gauge_lon', 'gauge_elev', 'area', 'elev_10', 'elev_50', 'elev_90']

static_attributes = climatic_attributes + \
    human_influence_attributes + \
    hydrogeology_attributes + \
    hydrologic_attributes + \
    landcover_attributes + \
    soil_attributes + \
    topographic_attributes

dynamic_inputs = ['P', 'T', 'EA', 'AMV']
# static_attributes = ['gauge_lat', 'gauge_lon', 'p_mean', 'pet_mean', 'area'] #, 'q_mean']#, 'frac_snow', 'high_prec_freq', 'high_prec_dur', 'low_prec_freq', 'low_prec_dur']

yaml = YAML() #typ = 'safe')
cfg = yaml.load(Path('resources/nh-config-template.yml'))
cfg['experiment_name'] = 'cudalstm_' + str(n_stations) + '_basins_' + str(aggr_period)
cfg['run_dir'] = rundir
cfg['train_basin_file'] = os.path.join(outputdir, "basins.txt")
cfg['validation_basin_file'] = os.path.join(outputdir, "basins.txt")
cfg['test_basin_file'] = os.path.join(outputdir, "basins.txt")
cfg['train_start_date'] = ["01/12/1961", "01/12/1992"] #"01/12/1961"
cfg['train_end_date'] = ["01/12/1990", "01/12/2006"] #"01/12/1990", #"01/12/2006"
cfg['validation_start_date'] = "01/12/1991" #"01/12/1961"
cfg['validation_end_date'] = "01/12/1991"
cfg['test_start_date'] = "01/12/1961"
cfg['test_end_date']= "01/12/2006"
cfg['device'] = "cpu" # "cuda:0"
cfg['validate_every'] = int(3)
cfg['validate_n_random_basins'] = int(1)
cfg['metrics'] = ["MSE"]
cfg['save_validation_results'] = True
cfg['model'] = "cudalstm" #"cudalstm" # "ealstm"
cfg['head'] = "regression"
cfg['output_activation'] = "linear"
cfg['hidden_size'] = int(64) #int(128) #int(20) #int(64) #int(20)
cfg['initial_forget_bias'] = int(3)
cfg['output_dropout'] = 0.4
cfg['optimizer'] = "Adam"
cfg['loss'] = "MSE"
cfg['learning_rate'] = {0: 1e-2, 30: 5e-3, 40: 1e-3}
cfg['batch_size'] = int(128) #int(256)
cfg['epochs'] = int(100) #int(50) #int(500) #int(150) #int(75)
cfg['clip_gradient_norm'] = int(1)
cfg['predict_last_n'] = int(1)
cfg['seq_length'] = int(2) #int(4) #int(8)
cfg['num_workers'] = int(8)
cfg['log_interval'] = int(5)
cfg['log_tensorboard'] = True
cfg['log_n_figures'] = int(0)
cfg['save_weights_every'] = int(1)
cfg['dataset'] = "generic"
cfg['data_dir'] = outputdir
cfg['dynamic_inputs'] = dynamic_inputs
cfg['target_variables'] = ["Q95"]
cfg['clip_targets_to_zero'] = ["Q95"]
cfg['static_attributes'] = static_attributes
conf_filename = os.path.join(outputdir, 'basins.yml')
with open(conf_filename, 'wb') as f:
    yaml.dump(cfg, f)

# ##################################### #
# 4 - Basins list                       #
# ##################################### #

basin_filename = os.path.join(outputdir, 'basins.txt')
with open(basin_filename, 'w') as f:
    for stn in station_ids:
        f.write(str(stn))
        f.write('\n')
